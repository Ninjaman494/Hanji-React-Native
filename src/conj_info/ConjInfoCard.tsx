import React from "react";
import { Conjugation } from "../hooks/useConjugations";
import { Badge, Text, Title, useTheme } from "react-native-paper";
import { View } from "react-native";
import BaseCard from "../components/BaseCard";

export type ConjInfoCardProps = {
  conjugation: Conjugation;
};

const ConjInfoCard: React.FC<ConjInfoCardProps> = ({ conjugation }) => {
  const { colors, padding } = useTheme();

  const styles = {
    card: {
      marginHorizontal: padding?.horizontal,
    },
    badge: {
      paddingHorizontal: padding?.horizontal,
      marginRight: padding?.horizontal,
      backgroundColor: colors?.accent,
    },
    title: {
      fontSize: 18,
      color: colors?.primary,
      marginTop: 16,
      marginBottom: padding?.vertical,
    },
    conjugation: {
      fontSize: 24,
      marginTop: -16,
    },
    text: {
      fontSize: 20,
    },
    secondary: {
      fontSize: 18,
      color: "#808080",
    },
  };

  const reasons = conjugation?.reasons.map((reason) => {
    const index = reason.indexOf("(");
    return {
      type: reason.substring(0, index),
      details: reason.substring(index),
    };
  });

  return (
    <BaseCard
      title={conjugation?.name}
      style={styles.card}
      rightIcon={() => (
        <Badge size={30} visible={conjugation?.honorific} style={styles.badge}>
          Honorific
        </Badge>
      )}
    >
      <View style={{ paddingHorizontal: padding?.horizontal }}>
        <Text style={styles.conjugation}>{conjugation?.conjugation}</Text>
        <Title style={styles.title}>Pronunciation</Title>
        <Text style={styles.text}>{conjugation?.pronunciation}</Text>
        <Text style={styles.secondary}>{conjugation?.romanization}</Text>
        <Title style={styles.title}>Explanation</Title>
        {reasons.map((reason, index) => (
          <View style={{ marginBottom: padding?.vertical }} key={index}>
            <Text style={styles.text}>{reason.type}</Text>
            <Text style={styles.secondary}>{reason.details}</Text>
          </View>
        ))}
      </View>
    </BaseCard>
  );
};

export default ConjInfoCard;
