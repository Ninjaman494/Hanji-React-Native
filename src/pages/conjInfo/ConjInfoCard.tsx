import { BaseCard, HonorificBadge } from "components";
import { Conjugation } from "hooks/useConjugations";
import React from "react";
import { StyleSheet, View } from "react-native";
import { Text, Title, useTheme } from "react-native-paper";
import toTitleCase from "utils/toTitleCase";

export type ConjInfoCardProps = {
  conjugation: Conjugation;
};

const ConjInfoCard: React.FC<ConjInfoCardProps> = ({ conjugation }) => {
  const { colors, padding, textSizes } = useTheme();

  const styles = StyleSheet.create({
    card: {
      marginHorizontal: padding?.horizontal,
      marginBottom: 8,
    },
    title: {
      fontSize: textSizes?.cardTitle,
      fontWeight: "400",
      color: colors?.primary,
      marginTop: 16,
      marginBottom: padding?.vertical,
    },
    conjugation: {
      fontSize: 24,
      marginTop: -24,
    },
    text: {
      fontSize: 20,
    },
    secondary: {
      fontSize: 18,
      color: "#808080",
    },
  });

  const reasons = conjugation?.reasons.map((reason) => {
    const index = reason.indexOf("(");
    return {
      type: reason.substring(0, index),
      details: reason.substring(index),
    };
  });

  return (
    <BaseCard
      title={toTitleCase(conjugation?.name)}
      style={styles.card}
      rightIcon={() => <HonorificBadge visible={conjugation?.honorific} />}
    >
      <View style={{ paddingHorizontal: padding?.horizontal }}>
        <Text style={styles.conjugation} selectable>
          {conjugation?.conjugation}
        </Text>
        <Title style={styles.title}>Pronunciation</Title>
        <Text style={styles.text} selectable>
          {conjugation?.pronunciation}
        </Text>
        <Text style={styles.secondary} selectable>
          {conjugation?.romanization}
        </Text>
        <Title style={styles.title}>Explanation</Title>
        {reasons.map((reason, index) => (
          <View style={{ marginBottom: padding?.vertical }} key={index}>
            <Text style={styles.text} selectable>
              {reason.type}
            </Text>
            <Text style={styles.secondary} selectable>
              {reason.details}
            </Text>
          </View>
        ))}
      </View>
    </BaseCard>
  );
};

export default ConjInfoCard;
