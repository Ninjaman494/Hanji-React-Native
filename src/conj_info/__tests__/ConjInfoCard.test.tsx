import React from "react";
import { render } from "@testing-library/react-native";
import ConjInfoCard from "../ConjInfoCard";

describe("ConjInfoCard component", () => {
  const conjugation = {
    name: "name",
    conjugation: "conj",
    type: "declarative present",
    tense: "PRESENT" as "PRESENT",
    speechLevel: "INFORMAL_LOW" as "INFORMAL_LOW",
    honorific: false,
    pronunciation: "pronunciation",
    romanization: "romanization",
    reasons: ["reason 1 (details)", "reason 2 (details)", "reason 3 (details)"],
  };

  it("displays a regular conjugation", () => {
    const component = render(<ConjInfoCard conjugation={conjugation} />);

    expect(component).toMatchSnapshot();
  });

  it("displays an honorific conjugation", () => {
    const component = render(
      <ConjInfoCard conjugation={{ ...conjugation, honorific: true }} />
    );

    expect(component).toMatchSnapshot();
  });
});
