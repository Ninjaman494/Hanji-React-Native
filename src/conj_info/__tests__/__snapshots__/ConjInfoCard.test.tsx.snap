// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConjInfoCard component displays a regular conjugation 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "borderRadius": 4,
      "elevation": 5,
      "marginHorizontal": undefined,
      "shadowColor": "#000000",
      "shadowOffset": Object {
        "height": 4,
        "width": 0,
      },
      "shadowOpacity": 0.24,
      "shadowRadius": 5,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flexGrow": 1,
        "flexShrink": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingLeft": 16,
          },
          Object {
            "minHeight": 72,
          },
          Object {
            "marginTop": -8,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontFamily": "System",
                  "fontWeight": "500",
                  "textAlign": "left",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Array [
                    Object {
                      "minHeight": 30,
                    },
                    Object {
                      "marginBottom": 2,
                    },
                    Object {
                      "color": "#6200ee",
                      "fontSize": undefined,
                      "textTransform": "capitalize",
                    },
                  ],
                ],
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Array [
                    Object {
                      "minHeight": 30,
                    },
                    Object {
                      "marginBottom": 2,
                    },
                    Object {
                      "color": "#6200ee",
                      "fontSize": undefined,
                      "textTransform": "capitalize",
                    },
                  ],
                ],
              ],
            ]
          }
        >
          name
        </Text>
      </View>
      <View>
        <Text
          numberOfLines={1}
          style={
            Object {
              "alignSelf": "flex-end",
              "backgroundColor": "#03dac4",
              "borderRadius": 15,
              "color": "#000000",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "400",
              "height": 30,
              "lineHeight": 30,
              "marginRight": undefined,
              "minWidth": 30,
              "opacity": 0,
              "overflow": "hidden",
              "paddingHorizontal": undefined,
              "textAlign": "center",
              "textAlignVertical": "center",
            }
          }
        >
          Honorific
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "paddingHorizontal": 16,
          },
          Object {
            "paddingBottom": 16,
          },
          Object {
            "paddingHorizontal": 0,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "paddingHorizontal": undefined,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Object {
                "fontSize": 24,
                "marginTop": -16,
              },
            ]
          }
        >
          conj
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontFamily": "System",
                  "fontWeight": "500",
                  "textAlign": "left",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Object {
                    "color": "#6200ee",
                    "fontSize": 18,
                    "marginBottom": undefined,
                    "marginTop": 16,
                  },
                ],
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Object {
                    "color": "#6200ee",
                    "fontSize": 18,
                    "marginBottom": undefined,
                    "marginTop": 16,
                  },
                ],
              ],
            ]
          }
        >
          Pronunciation
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Object {
                "fontSize": 20,
              },
            ]
          }
        >
          pronunciation
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Object {
                "color": "#808080",
                "fontSize": 18,
              },
            ]
          }
        >
          romanization
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontFamily": "System",
                  "fontWeight": "500",
                  "textAlign": "left",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Object {
                    "color": "#6200ee",
                    "fontSize": 18,
                    "marginBottom": undefined,
                    "marginTop": 16,
                  },
                ],
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Object {
                    "color": "#6200ee",
                    "fontSize": 18,
                    "marginBottom": undefined,
                    "marginTop": 16,
                  },
                ],
              ],
            ]
          }
        >
          Explanation
        </Text>
        <View
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Object {
                  "fontSize": 20,
                },
              ]
            }
          >
            reason 1 
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Object {
                  "color": "#808080",
                  "fontSize": 18,
                },
              ]
            }
          >
            (details)
          </Text>
        </View>
        <View
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Object {
                  "fontSize": 20,
                },
              ]
            }
          >
            reason 2 
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Object {
                  "color": "#808080",
                  "fontSize": 18,
                },
              ]
            }
          >
            (details)
          </Text>
        </View>
        <View
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Object {
                  "fontSize": 20,
                },
              ]
            }
          >
            reason 3 
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Object {
                  "color": "#808080",
                  "fontSize": 18,
                },
              ]
            }
          >
            (details)
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`ConjInfoCard component displays an honorific conjugation 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "borderRadius": 4,
      "elevation": 5,
      "marginHorizontal": undefined,
      "shadowColor": "#000000",
      "shadowOffset": Object {
        "height": 4,
        "width": 0,
      },
      "shadowOpacity": 0.24,
      "shadowRadius": 5,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flexGrow": 1,
        "flexShrink": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingLeft": 16,
          },
          Object {
            "minHeight": 72,
          },
          Object {
            "marginTop": -8,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontFamily": "System",
                  "fontWeight": "500",
                  "textAlign": "left",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Array [
                    Object {
                      "minHeight": 30,
                    },
                    Object {
                      "marginBottom": 2,
                    },
                    Object {
                      "color": "#6200ee",
                      "fontSize": undefined,
                      "textTransform": "capitalize",
                    },
                  ],
                ],
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Array [
                    Object {
                      "minHeight": 30,
                    },
                    Object {
                      "marginBottom": 2,
                    },
                    Object {
                      "color": "#6200ee",
                      "fontSize": undefined,
                      "textTransform": "capitalize",
                    },
                  ],
                ],
              ],
            ]
          }
        >
          name
        </Text>
      </View>
      <View>
        <Text
          numberOfLines={1}
          style={
            Object {
              "alignSelf": "flex-end",
              "backgroundColor": "#03dac4",
              "borderRadius": 15,
              "color": "#000000",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "400",
              "height": 30,
              "lineHeight": 30,
              "marginRight": undefined,
              "minWidth": 30,
              "opacity": 1,
              "overflow": "hidden",
              "paddingHorizontal": undefined,
              "textAlign": "center",
              "textAlignVertical": "center",
            }
          }
        >
          Honorific
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "paddingHorizontal": 16,
          },
          Object {
            "paddingBottom": 16,
          },
          Object {
            "paddingHorizontal": 0,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "paddingHorizontal": undefined,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Object {
                "fontSize": 24,
                "marginTop": -16,
              },
            ]
          }
        >
          conj
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontFamily": "System",
                  "fontWeight": "500",
                  "textAlign": "left",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Object {
                    "color": "#6200ee",
                    "fontSize": 18,
                    "marginBottom": undefined,
                    "marginTop": 16,
                  },
                ],
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Object {
                    "color": "#6200ee",
                    "fontSize": 18,
                    "marginBottom": undefined,
                    "marginTop": 16,
                  },
                ],
              ],
            ]
          }
        >
          Pronunciation
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Object {
                "fontSize": 20,
              },
            ]
          }
        >
          pronunciation
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Object {
                "color": "#808080",
                "fontSize": 18,
              },
            ]
          }
        >
          romanization
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontFamily": "System",
                  "fontWeight": "500",
                  "textAlign": "left",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Object {
                    "color": "#6200ee",
                    "fontSize": 18,
                    "marginBottom": undefined,
                    "marginTop": 16,
                  },
                ],
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Object {
                    "color": "#6200ee",
                    "fontSize": 18,
                    "marginBottom": undefined,
                    "marginTop": 16,
                  },
                ],
              ],
            ]
          }
        >
          Explanation
        </Text>
        <View
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Object {
                  "fontSize": 20,
                },
              ]
            }
          >
            reason 1 
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Object {
                  "color": "#808080",
                  "fontSize": 18,
                },
              ]
            }
          >
            (details)
          </Text>
        </View>
        <View
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Object {
                  "fontSize": 20,
                },
              ]
            }
          >
            reason 2 
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Object {
                  "color": "#808080",
                  "fontSize": 18,
                },
              ]
            }
          >
            (details)
          </Text>
        </View>
        <View
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Object {
                  "fontSize": 20,
                },
              ]
            }
          >
            reason 3 
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Object {
                  "color": "#808080",
                  "fontSize": 18,
                },
              ]
            }
          >
            (details)
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
